ligo/math.mligo

```diff
@@ -39,6 +39,26 @@ let assert_nat (x, error_code : int * nat) : nat =
     | None -> (failwith error_code : nat)
     | Some n -> n
 
+(* `Bitwise.shift_right x y` is only defined for `y <= 256n`.
+    This function handles larger values of `y`.
+ *)
+let rec stepped_shift_right (x, y : nat * nat) : nat =
```

Comment:
:thinking: I wonder if basing the logic on `is_nat` would be more efficient, have you tried by any chance?

If not, this PR is pretty big already, I will check this myself later for the optimization issue.

Reply:
I'm not sure what you mean :open_mouth: 

> I will check this myself later for the optimization issue.

Sure, sounds good :smile:

Reply:
Oh I meant something like:
```ocaml
let rec stepped_shift_right (x, y : nat * nat) : nat =
    match is_nat (y - 256n) with
    | None -> Bitwise.shift_right x y
    | Some n ->
        let new_x = Bitwise.shift_right x 256n in
        stepped_shift_right (new_x, n)
```
which is to say, use `is_nat` once to do both the check and the subtraction (+ `abs`).

Will look at this later on, if it makes a difference it's probably not that big :slightly_smiling_face: